### PR TITLE
HEAL-83,HEAL-84,HEAL-85,HEAL-86

### PR DESCRIPTION
**Pull request by aishwaryaravi19**
HEAL-83 - Upload doctor credentials to S3 - Doctor credential Verification
HEAL-84 - Encrypt the file uploaded by doctor using AES and upload it to S3 - Doctor Credential Verification
HEAL-85 - Fetch the corresponding file key and retrieve records from S3 (file download for HCProvider using AES decryption) - Doctor Credential Verification
HEAL-86 - Provide the ability for the doctors to choose a healthcare provider that they need to send the verification request to
SES Integration for doctor credentials
Fixed File Download issue

### CODE REVIEW

File: client/src/components/Doctor/DoctorDashboard.tsx [hunk 1]
```diff
@@ -3,34 +3,75 @@ import React, { useContext, useState, useEffect } from 'react';
...
+  const {fetchHealthCareProviders, providers, handleRaiseRequest} = useContext(HealthContext);
+  const [hcProvider, setHCProvider] = useState<any>('');
+  const [providerId, setProviderId] = useState<any>('');
   const { fetchAllDoctors, doctorList } = useContext(HealthContext);
-
+  const s3 = new AWS.S3({
+    accessKeyId: process.env.REACT_APP_AWS_ACCESS_KEY_ID,
+    secretAccessKey: process.env.REACT_APP_AWS_SECRET_ACCESS_KEY,
+    region: process.env.REACT_APP_AWS_REGION
+  });
```
Comment (varunalla) [inline client/src/components/Doctor/DoctorDashboard.tsx]: these are senstive information, if leaked will have monetary imapct. should not be available on front end. please refactor to store them via node js api

File: client/src/providers/HealthProvider.tsx [hunk 2]
```diff
@@ -250,20 +254,45 @@ export const HealthProvider: React.FC<Props> = ({ children, ...props }) => {
...
+    try {
+      const web3modal = new Web3Modal();
+      const connection = await web3modal.connect();
+      const provider = new ethers.providers.Web3Provider(connection);
+      const signer = provider.getSigner();
+      const contract = await fetchContract(signer);
+      const hcProviders = await contract.getAllProviders();
+      setProviders(hcProviders);
+    } catch (error) {
+      setError('Error Loading Health Contract');
```
Comment (varunalla) [inline client/src/providers/HealthProvider.tsx]: please do not set error inside hooks. please throw the error to calling component.

File: client/src/components/HealthCareProvider/HCProviderDashboard.tsx [hunk 3]
```diff
@@ -2,12 +2,21 @@ import React, { FunctionComponent, useContext, useEffect } from 'react';
...
   const navigate = useNavigate();
   const { user, role, logout } = useContext(AuthContext);
   const { handleApproveRequest, handleRejectRequest, verificationRequests, fetchRequests } =
     useContext(HealthContext);
+  const s3 = new AWS.S3({
+    accessKeyId: process.env.REACT_APP_AWS_ACCESS_KEY_ID,
+    secretAccessKey: process.env.REACT_APP_AWS_SECRET_ACCESS_KEY,
+    region: process.env.REACT_APP_AWS_REGION
+  });
+
```
Comment (varunalla) [inline client/src/components/HealthCareProvider/HCProviderDashboard.tsx]: Same Recommendation. this should not exist here. this is security problem.

File: client/src/components/Doctor/DoctorDashboard.tsx [hunk 4]
```diff
@@ -3,34 +3,75 @@ import React, { useContext, useState, useEffect } from 'react';
...
+        });
+      }
     } catch (err) {
       console.log(err);
     }
   };
 
+  useEffect(() => {
+    (() => {
+      fetchHealthCareProviders?.();
```
Comment (varunalla) [inline client/src/components/Doctor/DoctorDashboard.tsx]: should you not await this?

File: client/package-lock.json
Comment (varunalla) [inline client/package-lock.json:1]: delete this file

File: reencryptionproxy/server.py [hunk 5]
```diff
@@ -3,7 +3,7 @@
 from flask import Flask, request, jsonify
 from umbral import Capsule, PublicKey, SecretKey, Signer, VerifiedCapsuleFrag, VerifiedKeyFrag, decrypt_original, encrypt, generate_kfrags, reencrypt, decrypt_reencrypted
 from Crypto.Cipher import AES
-from Crypto.Random import get_random_bytes
+#from Crypto.Random import get_random_bytes
```
Comment (varunalla) [inline reencryptionproxy/server.py]: why is this removed?

File: proxy-reencryption/proxyReencryptionRoutes.js [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    res.status(500).send(err);
+  }
+});
+
+router.post('/decryptReencrypt', async (req, res) => {
+  const { private_key, public_key, capsule, ciphertext, cfrags } = req.body;
+
+  try {
+    // const cfrags_verified = cfrags.map(cfrag => VerifiedCapsuleFrag.fromVerifiedBytes(Buffer.from(cfrag, 'base64')));
+    const response = await axios.post('http://127.0.0.1:5001/decryptReencrypt', {
```
Comment (varunalla) [inline proxy-reencryption/proxyReencryptionRoutes.js]: remove hardcoding of this url, use constants file.

File: credentials/credentials-route.js [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+    const filenameWithoutExtension = path.basename(fileName, extension);
+    const key = req.body.key;
+  
+    if (!fileName) {
+      return res.status(400).json({ error: 'File name is required' });
+    }
+ 
+    try {
+      const params = {
+        Bucket: process.env.REACT_APP_BUCKET_NAME,
```
Comment (varunalla) [inline credentials/credentials-route.js:87]: is the bucket shared across all user personas?

File: credentials/credentials-route.js [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+});
+
+// Create a Multer instance
+const upload = multer({ storage });
+
+app.post('/upload', upload.single('file'), (req, res) => {
+  
+    try {
+      const file = req.file;
+      const fileName = req.body.fileName;
```
Comment (varunalla) [inline credentials/credentials-route.js:38]: what if two users upload the same file? where is the metadata stored for mapping between files and users?

**Review by varunalla — CHANGES_REQUESTED** [review]

File: credentials/credentials-route.js [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+
+      if (!file) {
+        return res.status(400).json({ error: 'No file provided' });
+      }
+
+      encryptor.encryptFile(req.file.path, filenameWithoutExtension+'-'+key+'.dat', key, function(err) {
+        const encryptedFile = fs.readFileSync('./'+filenameWithoutExtension+'-'+key+'.dat');
+        const params = {
+          Bucket: process.env.REACT_APP_BUCKET_NAME,
+          Key: filenameWithoutExtension+key,
```
Comment (varunalla) [inline credentials/credentials-route.js:52]: how do we make sure this key is unique?

**Review by varunalla — CHANGES_REQUESTED** [review]

File: credentials/credentials-route.js (hunk 7 above)
Comment (aishwaryaravi19) [reply credentials/credentials-route.js:87]: no, this is only for doctor credentials

File: reencryptionproxy/server.py (hunk 5 above)
Comment (aishwaryaravi19) [reply reencryptionproxy/server.py]: This is an unused import.

**Review by varunalla — APPROVED** [review]

File: credentials/credentials-route.js (hunk 9 above)
Comment (aishwaryaravi19) [reply credentials/credentials-route.js:52]: Made the key unique by generating a unique key using nanoid lib

File: credentials/credentials-route.js (hunk 8 above)
Comment (aishwaryaravi19) [reply credentials/credentials-route.js:38]: since a unique key is used to store the files in the bucket and for encryption, we will not have any overrides here
